packages/frint/src/App.ts

```diff
@@ -0,0 +1,410 @@
+  container: string;
+  store: string;
+  app: string;
+  parentApp: string;
+  rootApp: string;
+  region: string;
+}
+
+export interface RegisterAppOptions {
+  name?: string;
+  multi?: boolean;
+  regions?: string[];
+  initialize?: () => void;
+}
+
+export interface FrintProvider extends Provider {
+  cascade?: boolean;
+  scoped?: boolean;
+}
+
+const defaultProviderNames = {
+  component: 'component',
+  container: 'container',
+  store: 'store',
+  app: 'app',
+  parentApp: 'parentApp',
+  rootApp: 'rootApp',
+  region: 'region',
+};
+
+export interface AppRegistration {
+  appClass: AppClass;
+  name: string;
+  regions: any[];
+  instances: { [name: string]: App };
+  multi: boolean;
+  initialize?: () => void;
+}
+
+export interface AppOptions {
+  name?: string;
+  parentApp?: App;
+  providers?: FrintProvider[];
+  providerNames?: ProviderNames;
+  initialize?: () => void;
+  beforeDestroy?: () => void;
+}
+
+export interface AppClass {
+  frintAppName?: string;
+  new(opts?: AppOptions): App;
+}
+
+export class App {
+  public container: Container;
+  private options: AppOptions;
+  private _appsCollection: AppRegistration[];
+  private _apps$: BehaviorSubject<AppRegistration[]>;
+
+  constructor(opts: AppOptions) {
+    this.options = {
+      name: null,
+      parentApp: null,
+      providers: [],
+
+      providerNames: defaultProviderNames,
+
+      // lifecycle callbacks
+      // tslint:disable-next-line:no-empty
+      initialize: () => {},
+      // tslint:disable-next-line:no-empty
+      beforeDestroy: () => {},
+
+      // override
+      ...opts,
+    };
+
+    // errors
+    if (!this.options.name) {
+      throw new Error('Must provide `name` in options');
+    }
+
+    // children - create Observable if root
+    this._appsCollection = [];
+    this._apps$ = new BehaviorSubject(this._appsCollection);
+
+    // container
+    const container = createContainer([
+      { name: this.options.providerNames.app, useDefinedValue: this },
+      { name: this.options.providerNames.parentApp, useDefinedValue: this.getParentApp() },
+      { name: this.options.providerNames.rootApp, useDefinedValue: this.getRootApp() },
+    ], {
+      containerName: this.options.providerNames.container,
+    });
+
+    this.container = resolveContainer(container);
+
+    // root app's providers
+    this._registerRootProviders();
+
+    // self providers
+    this.options.providers.forEach((provider) => {
+      this.container.register(provider);
+    });
+
+    this.options.initialize.bind(this)();
+  }
+
+  public getContainer() {
+    return this.container;
+  }
+
+  public getRootApp() {
+    const parents = this.getParentApps();
+
+    if (parents.length === 0) {
+      return this;
+    }
+
+    return parents.pop();
+  }
+
+  public getParentApp() {
+    return this.options[this.options.providerNames.parentApp] as App || null;
+  }
+
+  public getParentApps() {
+    function findParents(app: App, parents: App[] = []): App[] {
```

Comment:
if there is no usage of `this` inside this function, please feel free to move this function out of this method. can stay in the same file as a private function that is not exported.

Reply:
Isn't it nicer if it stays inside the function? I mean this is a local helper just to be able to do recursion, right? So it's really local to this function.

Reply:
> for every call, function B will be defined again and again, and then called

Huh, I'm very curious about this one :)

Reply:
I tried to test it locally with Node, and I had to increase the number of iterations to 100,000,000 until I noticed a measurable difference, so in this case the overhead is neglibible (especially since the local `findParents` is created only once per `getParentApps` call.  
In terms of memory overhead—according to what I found—a local function takes up approximately as much memory as an empty object (like `let foo = {};`).

So I wouldn't be worried about overhead here at all, I'd just choose whichever is more readable. If you prefer the other one, I can change it.

Reply:
👍